graph/product/product.go

```diff
@@ -287,3 +301,45 @@ func indexOf(nodes []graph.Node) map[int64]int {
 	}
 	return idx
 }
+
+// complement provides the complement of a graph.
+type complement struct {
+	graph.Graph
+}
+
+func (g complement) From(id int64) graph.Nodes {
+	var cmpl []graph.Node
+	nodes := g.Nodes()
+	for nodes.Next() {
+		v := nodes.Node()
+		vid := v.ID()
+		if g.Graph.Edge(id, vid) == nil && id != vid {
+			cmpl = append(cmpl, v)
+		}
+	}
+	return iterator.NewOrderedNodes(cmpl)
+}
+
+func (g complement) Edge(uid, vid int64) graph.Edge {
+	if g.Graph.Edge(uid, vid) != nil || uid == vid {
+		return nil
+	}
+	u := g.Node(uid)
+	v := g.Node(vid)
+	if u == nil || v == nil {
+		return nil
+	}
+	return shadow{F: u, T: v}
+}
+
+func (g complement) HasEdgeBetween(xid, yid int64) bool {
+	return xid != yid &&
+		g.Node(xid) != nil && g.Node(yid) != nil &&
+		!g.Graph.HasEdgeBetween(xid, yid)
+}
+
+type shadow struct{ F, T graph.Node }
+
+func (e shadow) From() graph.Node         { return e.F }
+func (e shadow) To() graph.Node           { return e.T }
+func (e shadow) ReversedEdge() graph.Edge { return shadow{F: e.T, T: e.F} }
```

Comment:
This may make a worthwhile addition elsewhere though. Probably in graph itself, along with other operations like union, intersection and difference.